src/node_sets.cpp

```diff
@@ -0,0 +1,320 @@
+        throw SonataError("Compound node_set recursion depth exceeded");
+    }
+
+    for (auto const& it : compound_rules) {
+        for (auto const& name : it.second) {
+            if (compound_rules.count(name) > 0) {
+                check_compound(node_sets, compound_rules, depth + 1);
+            } else if (node_sets.count(name) == 0) {
+                throw SonataError(fmt::format("Missing '{}' from node_sets", name));
+            }
+        }
+    }
+}
+
+void parse_compound(const json& j, std::map<std::string, NodeSetRules>& node_sets) {
+    std::map<std::string, CompoundTargets> compound_rules;
+    for (auto& el : j.items()) {
+        if (el.value().is_array()) {
+            CompoundTargets targets;
+            for (const auto& name : el.value()) {
+                if (!name.is_string()) {
+                    throw SonataError("All compound elements must be strings");
+                }
+
+                targets.emplace_back(name);
+            }
+            compound_rules[el.key()] = targets;
+        }
+    }
+
+    check_compound(node_sets, compound_rules, 0);
+
+    for (const auto& rule : compound_rules) {
+        NodeSetRules rules;
+        rules.emplace_back(new NodeSetCompoundRule(rule.first, rule.second));
+        node_sets.emplace(rule.first, std::move(rules));
+    }
+}
+}  // anonymous namespace
+
+NodeSets::NodeSets(const std::string& content) {
+    json j = json::parse(content);
+    if (!j.is_object()) {
+        throw SonataError("Top level node_set must be an object");
+    }
+
+    // Need to two pass parsing the json so that compound lookup can rely
+    // on all the basic rules existing
+    parse_basic(j, node_sets_);
+    parse_compound(j, node_sets_);
+}
+
+Selection NodeSets::materialize(const std::string& name, const NodePopulation& population) const {
+    Selection ret = population.selectAll();
+
+    const auto& node_set = node_sets_.find(name);
+    for (const auto& ns : node_set->second) {
+        Selection a = ns->materialize(*this, population);
+        ret = ret & a;
+    }
+    return ret;
+}
+
+
+std::set<std::string> NodeSets::names() const {
+    std::set<std::string> ret;
+    std::transform(begin(node_sets_),
+                   end(node_sets_),
+                   std::inserter(ret, ret.begin()),
+                   [](decltype(node_sets_)::value_type const& pair) { return pair.first; });
+    return ret;
+}
+
+std::string NodeSets::toJSON() const {
+    auto replace_trailing_coma = [](std::string& s, char c) {
+        s.pop_back();
+        s.pop_back();
+        s.push_back(c);
+    };
+
+    std::string ret{"{\n"};
+    for (const auto& pair : node_sets_) {
+        ret += fmt::format(R"(  "{}": {{)", pair.first);
+        for (const auto& pred : pair.second) {
+            ret += pred->toJSON();
+            ret += ", ";
+        }
+        replace_trailing_coma(ret, ' ');
+        ret += "},\n";
+    }
+    replace_trailing_coma(ret, '\n');
+    ret += "}";
```

Comment:
JSON parsing is already included by an external library, why are we not using that instead of manually assembling all this?

Reply:
I was using it originally to serialize, but it got ugly (IMO).  The `toJSON` is mainly to help w/ debugging/testing, so it's valuable there, but I don't know how much usage it'll get 'in the real world'.

Reply:
Fair, the code is pretty simple, too. Maybe a comment on that?